### PR TITLE
Remove the >3D restriction on reshaping for dot_general

### PR DESCRIPTION
**Pull request by NatashaKnk**
Previously, the collapsing dims together (in case of multiple dims of the same type) would only happen if the input rank is larger than 3. This made the incorrect assumption that a 3d tensor is of a standard form (BxCxP or BxPxC), where in reality this could be the result of multiples of one dimension but none of another (BxBxC, for example). In those cases we still need to perform a reshape.

### CODE REVIEW

**Comment by antiagainst** [comment]
Note that we are building the StableHLO pipeline to replace the existing MHLO pipeline. Could you update the StableHLO pipeline with similar changes? @kuhar can give more pointers.

**Comment by kuhar** [comment]
> Note that we are building the StableHLO pipeline to replace the existing MHLO pipeline. Could you update the StableHLO pipeline with similar changes? @kuhar can give more pointers.

Given the github actions struggles today, we can first land this PR and I can incorporate the same changes into https://github.com/openxla/iree/pull/13519. But in the near-future, we should probably land changes in both and then only in the stablehlo pipeline.

**Comment by NatashaKnk** [comment]
We've discussed this with @MaheshRavishankar yesterday, and it seems like the right way to do this is to add the pass to the linalg lowering (which I'm guessing would address both). This is a temporary fix of sorts that addresses the failures.
I did see [13519](https://github.com/openxla/iree/pull/13519) adding the pass to StableHLO, so is the correct approach to add this change to that one as well?

**Review by kuhar — APPROVED** [review]

**Review by rsuderman — APPROVED** [review]

**Comment by MaheshRavishankar** [comment]
I am writing up an issue with details of what the next step are. I think part of that work will also including folding the stablehlo pattern and mhlo pattern into a single pattern at Linalg level. Sorry havent finished writing it yet, but will do so in a bit. I will tag both the StableHLO pr and this PR there.

File: compiler/src/iree/compiler/InputConversion/MHLO/test/mhlo_to_mhlo_preprocessing_canonicalize_dot_general.mlir [hunk 1]
```diff
@@ -15,3 +15,20 @@ func.func public @dot_general_2d(%arg0: tensor<4x3xf32> {mhlo.sharding = ""}, %a
...
+  // CHECK: %[[LHS_R:.+]] = mhlo.reshape %arg0 : (tensor<1x2x3xf32>) -> tensor<2x1x3xf32>
+  // CHECK: %[[RHS_R:.+]] = mhlo.reshape %[[RHS_T]] : (tensor<1x2x3x4xf32>) -> tensor<2x3x4xf32> 
+  // CHECK: "mhlo.dot_general"(%[[LHS_R]], %[[RHS_R]])
+  // CHECK-SAME: dot_dimension_numbers = #mhlo.dot<
+  // CHECK-SAME: lhs_batching_dimensions = [0]
+  // CHECK-SAME: rhs_batching_dimensions = [0]
+  // CHECK-SAME: lhs_contracting_dimensions = [2]
+  // CHECK-SAME: rhs_contracting_dimensions = [1]>
+  // CHECK-SAME: precision_config = [#mhlo<precision HIGHEST>, #mhlo<precision HIGHEST>]
+  return %0 : tensor<1x2x4xf32>
```
Comment (kuhar) [inline compiler/src/iree/compiler/InputConversion/MHLO/test/mhlo_to_mhlo_preprocessing_canonicalize_dot_general.mlir]: There seems to be a reshape of the `dot_general` result. Should we check it as well?

Comment (NatashaKnk) [reply compiler/src/iree/compiler/InputConversion/MHLO/test/mhlo_to_mhlo_preprocessing_canonicalize_dot_general.mlir]: Good point. Thanks!

**Pull request merged**